File: matlab_impl/+acoustics/PrecomputedRIRManager.m
classdef PrecomputedRIRManager < handle
    %PRECOMPUTEDRIRMANAGER Minimal manager for strict-equivalence runs.
    %
    % Provides the subset of RIRManager API required by ANC algorithms:
    % - ReferenceMicrophones / SecondarySpeakers / ErrorMicrophones dictionaries
    % - getSecondaryRIR(spkId, micId)

    properties
        ReferenceMicrophones dictionary
        SecondarySpeakers dictionary
        ErrorMicrophones dictionary
    end

    properties (Access = private)
        SecondaryRIRs dictionary
    end

    methods
        function obj = PrecomputedRIRManager(refIds, secIds, errIds, secRirs, secRirLengths)
            refIds = uint32(refIds(:).');
            secIds = uint32(secIds(:).');
            errIds = uint32(errIds(:).');

            if ndims(secRirs) ~= 3
                error('secRirs must be a 3D array [numSec, numErr, maxLen].');
            end

            if ~isequal(size(secRirLengths), [numel(secIds), numel(errIds)])
                error('secRirLengths must have size [numel(secIds), numel(errIds)].');
            end

            if size(secRirs, 1) ~= numel(secIds) || size(secRirs, 2) ~= numel(errIds)
                error('secRirs first two dimensions must match secIds and errIds lengths.');
            end

            obj.ReferenceMicrophones = configureDictionary('uint32', 'cell');
            obj.SecondarySpeakers = configureDictionary('uint32', 'cell');
            obj.ErrorMicrophones = configureDictionary('uint32', 'cell');
            obj.SecondaryRIRs = configureDictionary('string', 'cell');

            % Placeholder coordinates keep compatibility with existing algorithm code.
            for id = refIds
                obj.ReferenceMicrophones(id) = {[0 0 0]};
            end
            for id = secIds
                obj.SecondarySpeakers(id) = {[0 0 0]};
            end
            for id = errIds
                obj.ErrorMicrophones(id) = {[0 0 0]};
            end

            for i = 1:numel(secIds)
                for j = 1:numel(errIds)
                    nTaps = max(0, double(secRirLengths(i, j)));
                    if nTaps == 0
                        rir = 0;
                    else
                        rir = reshape(secRirs(i, j, 1:nTaps), 1, []);
                    end
                    key = "S" + string(secIds(i)) + "->M" + string(errIds(j));
                    obj.SecondaryRIRs(key) = {rir};
                end
            end
        end

        function h = getSecondaryRIR(obj, spkId, micId)
            key = "S" + string(spkId) + "->M" + string(micId);
            if ~isKey(obj.SecondaryRIRs, key)
                error('Secondary RIR for path (%s) does not exist.', key);
            end
            h = cell2mat(obj.SecondaryRIRs(key));
        end
    end
end

File: python_impl/py_anc/acoustics/precomputed_rir_manager.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Iterable, Tuple

import numpy as np


RIRKey = Tuple[int, int]


@dataclass
class PrecomputedRIRManager:
    """Lightweight manager for running ANC algorithms on precomputed paths."""

    reference_microphones: Dict[int, np.ndarray] = field(default_factory=dict)
    secondary_speakers: Dict[int, np.ndarray] = field(default_factory=dict)
    error_microphones: Dict[int, np.ndarray] = field(default_factory=dict)
    _secondary_rirs: Dict[RIRKey, np.ndarray] = field(default_factory=dict)

    @classmethod
    def from_padded_arrays(
        cls,
        ref_ids: Iterable[int],
        sec_ids: Iterable[int],
        err_ids: Iterable[int],
        sec_rirs: np.ndarray,
        sec_rir_lengths: np.ndarray,
    ) -> "PrecomputedRIRManager":
        manager = cls()

        ref_ids = [int(v) for v in np.asarray(ref_ids).reshape(-1)]
        sec_ids = [int(v) for v in np.asarray(sec_ids).reshape(-1)]
        err_ids = [int(v) for v in np.asarray(err_ids).reshape(-1)]

        sec_rirs = np.asarray(sec_rirs, dtype=float)
        sec_rir_lengths = np.asarray(sec_rir_lengths, dtype=int)

        if sec_rirs.ndim != 3:
            raise ValueError(f"sec_rirs must be 3D, got shape {sec_rirs.shape}.")
        if sec_rir_lengths.shape != sec_rirs.shape[:2]:
            raise ValueError(
                "sec_rir_lengths shape must match first two dimensions of sec_rirs, "
                f"got {sec_rir_lengths.shape} vs {sec_rirs.shape[:2]}."
            )
        if sec_rirs.shape[0] != len(sec_ids) or sec_rirs.shape[1] != len(err_ids):
            raise ValueError(
                "sec_rirs dimensions must match sec_ids/err_ids lengths, "
                f"got sec_rirs={sec_rirs.shape}, sec_ids={len(sec_ids)}, err_ids={len(err_ids)}."
            )

        # Positions are not used by algorithms in strict mode; placeholders keep API compatibility.
        zero = np.zeros(3, dtype=float)
        manager.reference_microphones = {rid: zero.copy() for rid in ref_ids}
        manager.secondary_speakers = {sid: zero.copy() for sid in sec_ids}
        manager.error_microphones = {eid: zero.copy() for eid in err_ids}

        for i, sid in enumerate(sec_ids):
            for j, eid in enumerate(err_ids):
                n_taps = int(sec_rir_lengths[i, j])
                if n_taps <= 0:
                    manager._secondary_rirs[(sid, eid)] = np.zeros(1, dtype=float)
                    continue
                manager._secondary_rirs[(sid, eid)] = sec_rirs[i, j, :n_taps].astype(float, copy=True)

        return manager

    def get_secondary_rir(self, spk_id: int, mic_id: int) -> np.ndarray:
        key = (int(spk_id), int(mic_id))
        if key not in self._secondary_rirs:
            raise KeyError(f"Secondary RIR for path {key} does not exist.")
        return self._secondary_rirs[key]
